src/include/migraphx/op/nonmaxsuppression.hpp

```diff
@@ -0,0 +1,259 @@
+            intersection_x_max = std::min(x1_max, x2_max);
+            if(intersection_x_max <= intersection_x_min)
+                return false;
+
+            y1_min = box1[1] - box1_height_half;
+            y1_max = box1[1] + box1_height_half;
+            y2_min = box2[1] - box2_height_half;
+            y2_max = box2[1] + box2_height_half;
+
+            intersection_y_min = std::max(y1_min, y2_min);
+            intersection_y_max = std::min(y1_max, y2_max);
+            if(intersection_y_max <= intersection_y_min)
+                return false;
+        }
+
+        const float intersection_area =
+            (intersection_x_max - intersection_x_min) * (intersection_y_max - intersection_y_min);
+
+        if(intersection_area <= .0f)
+        {
+            return false;
+        }
+
+        const float area1      = (x1_max - x1_min) * (y1_max - y1_min);
+        const float area2      = (x2_max - x2_min) * (y2_max - y2_min);
+        const float union_area = area1 + area2 - intersection_area;
+
+        if(area1 <= .0f || area2 <= .0f || union_area <= .0f)
+        {
+            return false;
+        }
+
+        const float intersection_over_union = intersection_area / union_area;
+
+        return intersection_over_union > iou_threshold;
+    }
+
+    argument compute(const shape& output_shape, std::vector<argument> args) const
+    {
+        argument result{output_shape};
+
+        result.visit([&](auto out) { std::fill(out.begin(), out.end(), 0); });
+
+        int64_t max_output_boxes_per_class = 0;
+        float iou_threshold                = 0.0f;
+        float score_threshold              = 0.0f;
+
+        if(args.size() > 2)
+        {
+            max_output_boxes_per_class = args.at(2).at<int64_t>();
+        }
+        // max_output_boxes_per_class is 0, no output
+        if(max_output_boxes_per_class == 0)
+        {
+            return result;
+        }
+
+        if(args.size() > 3)
+        {
+            iou_threshold = args.at(3).at<float>();
+        }
+
+        if(args.size() > 4)
+        {
+            score_threshold = args.at(4).at<float>();
+        }
+
+        const auto& lens = args.at(1).get_shape().lens();
+        auto batch_num   = lens[0];
+        auto class_num   = lens[1];
+        auto box_num     = args.at(0).get_shape().lens()[1];
+
+        std::vector<box_info_ptr> selected_boxes_inside_class;
+        std::vector<int64_t> selected_indices;
+        selected_boxes_inside_class.reserve(output_shape.elements());
+
+        auto scores = make_view<float>(args.at(1).get_shape(), args.at(1).cast<float>());
+        shape comp_s{shape::float_type, {batch_num, class_num}};
+        shape_for_each(comp_s, [&](auto idx) {
+            auto bidx = idx[0];
+            auto cidx = idx[1];
+
+            std::size_t score_offset = (bidx * class_num + cidx) * box_num;
+            const float* batch_boxes = args.at(0).cast<float>() + bidx * box_num * 4;
```

Comment:
Use an iterator instead of a raw pointer.

Reply:
add a function batch_box()